modules/core/portal-web/src/main/java/com/liferay/portal/web/internal/session/replication/SessionReplicationHttpSessionWrapper.java

```diff
@@ -0,0 +1,135 @@
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.web.internal.session.replication;
+
+import com.liferay.petra.io.Deserializer;
+import com.liferay.petra.io.Serializer;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.servlet.HttpSessionWrapper;
+import com.liferay.portal.kernel.util.JavaDetector;
+
+import java.io.Serializable;
+
+import java.nio.ByteBuffer;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.logging.Logger;
+
+import javax.servlet.http.HttpSession;
+
+/**
+ * @author Dante Wang
+ */
+public class SessionReplicationHttpSessionWrapper extends HttpSessionWrapper {
+
+	public SessionReplicationHttpSessionWrapper(HttpSession session) {
+		super(session);
+	}
+
+	@Override
+	public Object getAttribute(String name) {
+		Object value = super.getAttribute(
+			_SERIALIZED_ATTRIBUTE_PREFIX.concat(name));
+
+		if (value == null) {
+			return super.getAttribute(name);
+		}
+
+		Deserializer deserializer = new Deserializer(
+			ByteBuffer.wrap((byte[])value));
+
+		try {
+			return deserializer.readObject();
+		}
+		catch (ClassNotFoundException classNotFoundException) {
+			_log.error("Unable to deserialize object", classNotFoundException);
+
+			return null;
+		}
+	}
+
+	@Override
+	public Enumeration<String> getAttributeNames() {
+		Enumeration<String> attributeNameEnumeration =
+			super.getAttributeNames();
+
+		List<String> attributeNames = new ArrayList<>();
+
+		while (attributeNameEnumeration.hasMoreElements()) {
+			String attributeName = attributeNameEnumeration.nextElement();
+
+			if (attributeName.startsWith(_SERIALIZED_ATTRIBUTE_PREFIX)) {
+				attributeName = attributeName.substring(
+					_SERIALIZED_ATTRIBUTE_PREFIX.length());
+			}
+
+			attributeNames.add(attributeName);
+		}
+
+		return Collections.enumeration(attributeNames);
+	}
+
+	@Override
+	public void removeAttribute(String name) {
+		super.removeAttribute(name);
+
+		super.removeAttribute(_SERIALIZED_ATTRIBUTE_PREFIX.concat(name));
+	}
+
+	@Override
+	public void setAttribute(String name, Object value) {
+		if (!(value instanceof Serializable) ||
+			_isSafeClass(value.getClass())) {
+
+			super.setAttribute(name, value);
+
+			return;
+		}
+
+		Serializer serializer = new Serializer();
+
+		serializer.writeObject((Serializable)value);
+
+		ByteBuffer byteBuffer = serializer.toByteBuffer();
+
+		super.setAttribute(
+			_SERIALIZED_ATTRIBUTE_PREFIX.concat(name), byteBuffer.array());
+	}
+
+	private boolean _isSafeClass(Class<?> clazz) {
+		ClassLoader classLoader = clazz.getClassLoader();
+
+		if ((classLoader == String.class.getClassLoader()) ||
```

Comment:
@tinatian I keep this to use `== String.class.getClassLoader` instead of `== null` because the Java Doc of `Class.getClassLoader()` says "some (Java) implementations may use null to represent the bootstrap class loader" -- Oracle JDK does return null, but I'm not sure about others.